### ISSUE TITLE
7.12.0 can't add properties to propType 

### ISSUE THREAD

**Issue by cyrildurand** [yannickcr/eslint-plugin-react#2122]
Since v7.12.0 it is not possible to alter propTypes anymore 

Let's consider this code 
```jsx
        const personPropTypes = {};
        personPropTypes.name = PropTypes.string.isRequired;
        var Hello = createReactClass({
          propTypes: personPropTypes,,
          render: function() {,
            return <div>Hello {this.props.name}</div>;,
          },
        });
```
With 7.11.1 this code validates without error but since 7.12.0 we have a 
> name is missing in props validation

I need to alter the propTypes object to support hierarchy. ie : 
```jsx
    const menuItemPropTypes = {
      id: PropTypes.string.isRequired,
      title: PropTypes.string.isRequired,
      path: PropTypes.string,
    };
    menuItemPropTypes.menuItems = PropTypes.arrayOf(PropTypes.shape(menuItemPropTypes));
```

**Comment by ljharb** [comment]
That’s not actually necessary - try this:
```js
    const menuItemPropTypes = {
      id: PropTypes.string.isRequired,
      title: PropTypes.string.isRequired,
      path: PropTypes.string,
      menuItems(...args) {
        return PropTypes.arrayOf(PropTypes.shape(menuItemsPropTypes))(...args);
      },
    };
```

Mutation after the fact is almost never required.

**Comment by cyrildurand** [comment]
Thanks for this solution but in that case, I have an eslint error 'menuItemsPropTypes' is not defined.' 

I can ignore this error message but is there any other way of doing it ?

**Comment by ljharb** [comment]
Ah, true. I’ve updated my post.

**Comment by cyrildurand** [comment]
After further investigation it looks like I can reproduce the issue with this minimal code 

    const fooPropTypes = {
      firstName: PropTypes.string.isRequired,
    };
    fooPropTypes.lastName = PropTypes.string.isRequired;

    export class FooComponent extends React.Component {
      static propTypes = fooPropTypes;

      render() {
        const { firstName, lastName, invalid } = this.props;
        const o = { firstName, lastName, invalid };
        return <span {...o} />;
      }
    }

With 7.11.0 the spread operator (`{...o}`) acts like it disabled eslint for my class : I have no error for `lastName` nor `invalid`. If I remove it I do have `'lastName' is missing in props validation` and `'invalid' is missing in props validation' errors. 

After updating to 7.12.0 (and restarting VS Code) I do have the errors with the spread operator. 

I found this stackoverflow question [Spread operator on a component disables the react/prop-types rule on eslint-config-airbnb](https://stackoverflow.com/questions/52521116/spread-operator-on-a-component-disables-the-react-prop-types-rule-on-eslint-conf) with a similar issue. 

It seems to be related to #1939

Thanks for the help
